### ISSUE TITLE
mat-tree should have a filtering feature

### ISSUE THREAD

**Issue by ajaysattikar** [angular/material2#11505]
#### Feature request, or proposal:
Filtering through tree nodes is a very useful feature, especially to reduce the size of tree and finding out the desired node(s) quickly.

#### What is the expected behavior?
There should be an input box above mat-tree. As you enter the text, it will filter through the tree and renders the tree based on the content in the input box.

#### What is the current behavior?
Filtering feature is not available in mat-tree component.

**Comment by mehfatih** [comment]
I tried to create a solution myself on the link below. It can help those who need it.

[stackblitz sample](https://stackblitz.com/edit/angular-yb37gh?file=app%2Ftree-checklist-example.html)

**Comment by rojagadeesh** [comment]
> I tried to create a solution myself on the link below. It can help those who need it.
> 
> [stackblitz sample](https://stackblitz.com/edit/angular-yb37gh?file=app%2Ftree-checklist-example.html)

@mehfatih - You applied filter successfully applied to Mat-tree and works fine.  But when I click checkbox and remove the filtertext, then the checked values are missing.

**Comment by hugocalheira** [comment]
The filter could work like this one: https://angular2-tree.readme.io/docs/filtering

**Comment by rojagadeesh** [comment]
No, I didn't found any solution how to apply filter to mat-tree

**Comment by NaveenMysuru** [comment]
solution given by Mr.mehfatih works well for linear tree data. but it fails to track node path while searching in nested tree data like in official documentation examples.

**Comment by AbhishekChugh** [comment]
+1 this would be really useful

**Comment by shantd9** [comment]
Indeed, this would be very useful

**Comment by Hartaxfr** [comment]
Need it too

**Comment by wilhelmmi** [comment]
No solutions so far? Struggling with this issue too...

**Comment by shirlebel** [comment]
I wrote this recursive function that implements search in nested tree and it works:

`  //gets a tree and returns a new tree includes only nodes relevant to the "searchedText"
  //will return a tree that includes the searched nodes, all searched nodes' descendants, and the direct path of the forefathers
  private filterTreeRecursion(searchedText: string, node: Node) :Node{
    let newParentNode: Node = {
      children: [],
      item: node.item,
      id: node.id,
      pid: node.pid,
      showIcons: node.showIcons,
      data: node.data,
      character: node.character,
    }
    if (node.item.toLowerCase().includes(searchedText.toLowerCase())) { //check if node's name includes the searched string
      return node;
    }

    if (node.children && node.children.length > 0) {
      //going through all node's children
      for (let i = 0; i < node.children.length; i++) {
        let resNode = this.filterTreeRecursion(searchedText, node.children[i]);
        if (resNode !== undefined) {
          newParentNode.children.push(resNode);
        }
      }
    }

    if (newParentNode.children.length > 0){
      return newParentNode;
    }
  }`

**Comment by shirlebel** [comment]
I wrote this recursive function that implements search in nested tree and it works.
It gets a root node, and returns a tree that includes the searched nodes, all searched nodes' descendants, and the direct path of the forefathers:

`private filterTreeRecursion(searchedText: string, node: Node) :Node{
    let newParentNode: Node = {
      children: [],
      item: node.item,
      id: node.id,
      pid: node.pid,
      showIcons: node.showIcons,
      data: node.data,
      character: node.character,
    }
    if (node.item.toLowerCase().includes(searchedText.toLowerCase())) { //check if node's name includes the searched string
      return node;
    }

    if (node.children && node.children.length > 0) {
      //going through all node's children
      for (let i = 0; i < node.children.length; i++) {
        let resNode = this.filterTreeRecursion(searchedText, node.children[i]);
        if (resNode !== undefined) {
          newParentNode.children.push(resNode);
        }
      }
    }

    if (newParentNode.children.length > 0){
      return newParentNode;
    }
  }`

**Comment by shirlebel** [comment]
I wrote this recursive function that implements search in nested tree and it works.
It gets a root node, and returns a tree that includes the searched nodes, all searched nodes' descendants, and the direct path of the forefathers:

`private filterTreeRecursion(searchedText: string, node: Node) :Node{
    let newParentNode: Node = {
      children: [],
      item: node.item,
      id: node.id,
      pid: node.pid,
      showIcons: node.showIcons,
      data: node.data,
      character: node.character,
    }
    if (node.item.toLowerCase().includes(searchedText.toLowerCase())) { //check if node's name includes the searched string
      return node;
    }

    if (node.children && node.children.length > 0) {
      //going through all node's children
      for (let i = 0; i < node.children.length; i++) {
        let resNode = this.filterTreeRecursion(searchedText, node.children[i]);
        if (resNode !== undefined) {
          newParentNode.children.push(resNode);
        }
      }
    }

    if (newParentNode.children.length > 0){
      return newParentNode;
    }
  }`

**Comment by shirlebel** [comment]
I wrote this recursive function that implements search in nested tree and it works.
It gets a root node, and returns a tree that includes the searched nodes, all searched nodes' descendants, and the direct path of the forefathers:

`private filterTreeRecursion(searchedText: string, node: Node) :Node{
    let newParentNode: Node = {
      children: [],
      item: node.item,
      id: node.id,
      pid: node.pid,
      showIcons: node.showIcons,
      data: node.data,
      character: node.character,
    }
    if (node.item.toLowerCase().includes(searchedText.toLowerCase())) { //check if node's name includes the searched string
      return node;
    }

    if (node.children && node.children.length > 0) {
      //going through all node's children
      for (let i = 0; i < node.children.length; i++) {
        let resNode = this.filterTreeRecursion(searchedText, node.children[i]);
        if (resNode !== undefined) {
          newParentNode.children.push(resNode);
        }
      }
    }

    if (newParentNode.children.length > 0){
      return newParentNode;
    }
  }`

**Comment by shirlebel** [comment]
I wrote this recursive function that implements search in nested tree and it works.
It gets a root node, and returns a tree that includes the searched nodes, all searched nodes' descendants, and the direct path of the forefathers:

`
  //gets a tree root and returns a NEW tree root includes only nodes relevant to the "searchedText"
  //will return a tree that includes the searched nodes, all searched nodes' descendants, and the direct path of the forefathers
  private filterTreeRecursion(searchedText: string, node: Node): Node {
    // if node's name includes the searched string - returns it
    if (node.item.toLowerCase().includes(searchedText.toLowerCase())) { 
      return node;
    }

    //if not - continue to check for it's children
    if (node.children && node.children.length > 0) {
      let newParentNode: Node = {
        children: [],
        item: node.item,
        id: node.id,
        pid: node.pid,
        showIcons: node.showIcons,
        data: node.data,
        character: node.character,
      }

      //going through all node's children
      for (let i = 0; i < node.children.length; i++) {
        let resNode = this.filterTreeRecursion(searchedText, node.children[i]);
        if (resNode !== undefined) {
          newParentNode.children.push(resNode); //if there is a result, it's asearched node or a forefather of the searched node - so we add it
        }
      }

      if (newParentNode.children.length > 0) {
        return newParentNode;
      }
    }
  }
`

**Comment by vickyadi3** [comment]
@shirlebel , Where can i put this function, because  when  I add my in this example it saying node.childern is undefined.

**Comment by shirlebel** [comment]
> @shirlebel , Where can i put this function, because when I add my in this example it saying node.childern is undefined.

notice that the function start with "private filterTreeRecursion(searchedText: string, node: Node)" (it's outside the marked square)
and check your node structure, if it dosn't have the children field you need to change it to your field

**Comment by Darius123** [comment]
i would love to have this feature

**Comment by Gagulo** [comment]
You may take a look [Here](https://stackblitz.com/edit/angular-tree-node?file=app%2Ftree-nested-overview-example.ts)

**Comment by JoaoPortella** [comment]
> You may take a look [Here](https://stackblitz.com/edit/angular-tree-node?file=app%2Ftree-nested-overview-example.ts)

WOW!

**Comment by miladSorour** [comment]
+1

**Comment by tmushayahama** [comment]
Anything on this?

**Comment by Amith-B** [comment]
I tried to visually filter the mat tree without modifying the datasource. Check this stackblitz link.

[https://stackblitz.com/edit/tree-with-checkbox-and-a-filter?file=src/app/tree-checklist-example.ts](url)

**Comment by wagnermaciel** [comment]
Hi, I'm trying to get a gauge on whether this is something we actually want to implement or if this is something that should be left to developers to handle on their own. I'm currently leaning towards the latter for the following reasons:

1. Even in this thread there have been a couple of pretty straightforward implementations
2. Filters seem opinionated and any one implementation would probably not satisfy all desired filtering strategies

Let me know your thoughts and if you agree or disagree

**Comment by viheag-98** [comment]
@wagnermaciel I'd really appreciate the implementation of this idea...

**Comment by LanderBeeuwsaert** [comment]
@wagnermaciel I would say that the implementation of the filter function is most probably different per implementation.
However, it would be nice that mat-tree has, like mat-table a property where you can "attach" your own filter function.

**Comment by wagnermaciel** [comment]
@LanderBeeuwsaert that would seem like the most practical approach to me as well

**Comment by tmushayahama** [comment]
A filter will be awesome. I think mat-tree is not complete and missing useful functionalities that a normal tree should have. I always resort to third party angular tree component.
